tsdb/db.go

```diff
@@ -638,6 +661,14 @@ func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs
 		return nil, err
 	}
 
+	// Set the min valid time for the ingested samples
+	// to be no lower than the maxt of the last block.
+	blocks := db.Blocks()
+	minValidTime := int64(math.MinInt64)
+	if len(blocks) > 0 {
+		minValidTime = blocks[len(blocks)-1].Meta().MaxTime
+	}
+	go truncateWAL(db.head, minValidTime)
```

Comment:
Is it possible that DB can be closed while truncation is still in progress? If yes, then the closing should wait for truncation.

Reply:
Additionally, we should initiate the truncation at the very end after the blocks reload, head init, etc is done - when we are sure that there are no errors opening the DB.